fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/sorting/FileSorter.java

```diff
@@ -0,0 +1,625 @@
+                secureRandom = null;
+                cipher = null;
+            }
+            fileHeader = RecordSortingProto.SortFileHeader.newBuilder()
+                    .setVersion(SORT_FILE_VERSION)
+                    .setNumberOfSections(0)
+                    .setNumberOfRecords(0);
+            headerStream.writeMessageNoTag(fileHeader.build());
+            sectionHeader = RecordSortingProto.SortSectionHeader.newBuilder()
+                    .setSectionNumber(0)
+                    .setStartRecordNumber(0)
+                    .setNumberOfRecords(0)
+                    .setNumberOfBytes(0);
+            writeSectionHeader();
+        }
+
+        public void next(@Nonnull byte[] key, @Nonnull byte[] value) throws IOException, GeneralSecurityException {
+            entryStream.writeByteArrayNoTag(key);
+            entryStream.writeByteArrayNoTag(value);
+            fileHeader.setNumberOfRecords(fileHeader.getNumberOfRecords() + 1);
+            sectionHeader.setNumberOfRecords(sectionHeader.getNumberOfRecords() + 1);
+            if (sectionHeader.getNumberOfRecords() >= recordsPerSection) {
+                rewriteSectionHeader();
+                fileHeader.setNumberOfSections(fileHeader.getNumberOfSections() + 1);
+                sectionHeader.setSectionNumber(fileHeader.getNumberOfSections());
+                sectionHeader.setStartRecordNumber(fileHeader.getNumberOfRecords());
+                sectionHeader.setNumberOfRecords(0).setNumberOfBytes(0);
+                writeSectionHeader();
+            }
+        }
+
+        public void finish() throws IOException {
+            rewriteSectionHeader();
+            final long fileLength = fileChannel.position();
+            fileChannel.position(0);
+            headerStream.writeMessageNoTag(fileHeader.build());
+            headerStream.flush();
+            fileChannel.position(fileLength);
+        }
+
+        @Override
+        public void close() throws IOException {
+            if (cipher != null) {
+                CipherPool.returnCipher(cipher);
+            }
+            fileStream.close();
+        }
+
+        void writeSectionHeader() throws IOException, GeneralSecurityException {
+            headerStream.flush();
+            sectionHeaderPosition = fileChannel.position();
+            if (cipher != null) {
+                initCipherEncrypt(cipher, encryptionKey, secureRandom, sectionHeader);
+            }
+            headerStream.writeMessageNoTag(sectionHeader.build());
+            headerStream.flush();
+            sectionRecordsPosition = fileChannel.position();
+            if (compress || cipher != null) {
+                headerStream.flush();
+                outputStream = wrapOutputStream(fileStream, cipher, compress);
+                entryStream = CodedOutputStream.newInstance(outputStream);
+            }
+        }
+
+        @SuppressWarnings("PMD.CompareObjectsWithEquals")
+        void rewriteSectionHeader() throws IOException {
```

Comment:
I actually think it would trade one kind of complexity for another. Assuming that the header file had positions in the record file, you'd still have all the skipping around there. And there would be two file handles to manage.